hs_file_types/forms.py

```diff
@@ -0,0 +1,282 @@
+            if schema_data:
+                return self._validate_json_schema(schema_string=schema_data, field_name='mi_json_schema_file')
+        return json_schema
+
+    def update_metadata(self, metadata):
+        metadata.version = self.cleaned_data['version']
+        metadata.website = self.cleaned_data['website']
+        metadata.code_repository = self.cleaned_data['code_repository']
+        metadata.release_date = self.cleaned_data['release_date']
+        metadata.operating_systems = self.cleaned_data['operating_systems']
+        metadata.programming_languages = self.cleaned_data['programming_languages']
+        metadata.is_dirty = True
+        metadata.save()
+        logical_file = metadata.logical_file
+
+        # if the user has uploaded json file for the metadata schema we will use the content
+        # of that file to populate the the mi_schema_json metadata field
+        if self.cleaned_data['mi_json_schema_file']:
+            # use the content from the uploaded json file to save in database
+            logical_file.mi_schema_json = self.cleaned_data['mi_json_schema_file']
+        else:
+            logical_file.mi_schema_json = self.cleaned_data['mi_json_schema']
+
+        logical_file.model_program_type = self.cleaned_data['mp_program_type']
+        logical_file.save()
+        # recreate ModelProgramResourceFileType objects
+        if self.cleaned_data["mp_file_types"]:
+            mp_file_types_dict = self.cleaned_data["mp_file_types"]
+            # delete all existing ModelProgramResourceFileType objects for this aggregation
+            metadata.mp_file_types.all().delete()
+            logical_file = metadata.logical_file
+            for res_file in logical_file.files.all():
+                if res_file.short_path in mp_file_types_dict:
+                    mp_file_type = mp_file_types_dict[res_file.short_path]
+                    # if mp_file_type is None it means user has removed model file type assignment for the res_file
+                    if mp_file_type is not None:
+                        ModelProgramResourceFileType.objects.create(file_type=mp_file_type, res_file=res_file,
+                                                                    mp_metadata=metadata)
+
+    def _validate_json_schema(self, schema_string, field_name):
+        """helper to validate json schema for model instance metadata
+        :param  schema_string: metadata json schema as a string - which needs to be validated
+        :param  field_name: form field name for which the json schema input is getting validated
+        :returns   a dict object containing the validated json schema
+        """
+        json_schema = dict()
+        is_schema_valid = True
+        try:
+            json_schema = json.loads(schema_string)
+        except ValueError as exp:
+            is_schema_valid = False
+            self.add_error(field_name, "Not a valid JSON string")
```

Comment:
Do a return here so you don't have to check json_schema and is_schema_valid below and drop the nested ifs by one.

Reply:
Even with a return here, still need to check json_schema is not an empty dict.